### ISSUE TITLE
Android PushDelegate OnEntry never called for push notification

### ISSUE THREAD

**Issue by GuidoNeele** [shinyorg/shiny#656]
ONLY active OSS maintainers of other projects OR ACTIVE project sponsors can ask questions here.  If you don't do either of these things - DO NOT FILE HERE OR THIS WILL BE CLOSED!
THERE IS NO SUPPORT ON PREVIEW/BETA VERSIONS!  

Give as much details as humanly possible if you want any sort of answer!

Enter Question Below (don't delete this line)
Hey Allan,

Got another question. I was testing the push notifications on Android (using Azure Notification Hub) and I can't get the `PushDelegate`.`OnEntry` method triggered when a notification is received via push and the app is in the background. I've debugged the code and the check for the `IntentNotificationKey` is always false when it's a push notification (see permalink below). The `IntentNotificationKey` is only set when the notification is shown via the Shiny `NotificationManager`. Is this a bug or expected behavior? 
https://github.com/shinyorg/shiny/blob/b33f1ac410861b47a2518e5e9c660f4a15e3c7ca/src/Shiny.Notifications/Platforms/Android/AndroidNotificationProcessor.cs#L34

**Comment by aritchie** [comment]
OnEntry isn't the method that is called, OnReceived is

**Comment by GuidoNeele** [comment]
Ok so I should be sending data messages only and handle them in OnReceived? If I want to notify the user show a local notification selecting the correct channel etc.

**Comment by GuidoNeele** [comment]
OK consider this flow:
1. App is in background
2. A notification like this is send:
```json
{
	"notification": {
		"title": "John Doe",
		"body": "This is the message"
	},
	"android": {
		"priority": "high",
		"notification": {
			"channel_id": "chat"
		}
	},
	"data": {
		"action": "chat_message_received",
		"Id": "1626964406847",
		"SequenceId": null,
		"SenderId": "8:acs:cb8501ac-57ab-4b95-a1d8-743989ac21ad_0000000b-4f09-5a53-7c4a-ad3a0d00f883",
		"ThreadId": "19:9vqlmpOaXX4XdbdOjH-OG_DxnsbOzoOC_IelKqktL8s1@thread.v2",
		"Type": "Text",
		"CreatedOn": "637625612068470000",
		"EditedOn": null,
		"DeletedOn": null,
		"Version": "1626964406847",
		"id": "dc8f3008-5557-4193-914c-c3cca2667ca0",
		"title": "John Doe",
		"message": "This is the message"
	}
}
```
3. Notification appears on the device.
4. User clicks the notification

After step 4 I thought OnEntry would trigger so I could redirect the user to another view.

**Comment by aritchie** [comment]
Ah ok - yes that should be the case.  Is this on a specific platform?

**Comment by GuidoNeele** [comment]
Yep only on Android. iOS looks totally different. It has to do with detecting if the intent is triggered by a Shiny local notification. For that the `IntentNotificationKey` is checked but this key doesn't exist on the push notifications. Only the notifications that are triggered by Shiny itself. This is because Shiny.Push Android is relying on the INotificationDelegate for OnEntry to trigger.
https://github.com/shinyorg/shiny/blob/a52b334a2e8167e3c33221e6182389f9352a1e5e/src/Shiny.Push/Platforms/Android/PushNotificationDelegate.cs

https://github.com/shinyorg/shiny/blob/b33f1ac410861b47a2518e5e9c660f4a15e3c7ca/src/Shiny.Notifications/Platforms/Android/AndroidNotificationProcessor.cs#L34

**Comment by aritchie** [comment]
Thanks - will look into it for 2.2

**Comment by GuidoNeele** [comment]
Thanks Allan!

**Comment by aritchie** [comment]
This didn't make it into 2.2 only because it isn't a simple fix.  It is a priority, but I needed more time to figure out to approach this.

**Comment by aritchie** [comment]
Please read the following.  Your thoughts are welcome here: https://github.com/shinyorg/shiny/discussions/726

**Comment by aritchie** [comment]
This has been an extremely hard issue to solve for all platforms especially Android.  I think I'm going to remove the ability to try and deal with actions from push altogether.  There will also be a requirement for android push to include the clickaction in the payload to link to the Shiny push intent on open which would allow Shiny to work with OnEntry.

If you need something with actions (buttons on the notification or any sort of customization), you need to send a silent push with the necessary details and use Shiny.Notifications to customize and deal with the OnEntry.  By doing this, I can simplify the necessary mechanics for push.

**Comment by GuidoNeele** [comment]
Sounds good to me!

**Comment by aspeckt112** [comment]
Hey @aritchie 

Firstly, I apologise if this isn't the correct way to do this. I'm not very experienced contributing to OSS, but I'm keen to learn.

I've run into this problem, specifically on Android when sending a notification message with an optional data payload. The equivalent message using APNS (I'm using native push) works fine on iOS - OnEntry is called as I would expect and I can do what I need to do there.

Regarding your comment above, 

> There will also be a requirement for android push to include the clickaction in the payload to link to the Shiny push intent on open which would allow Shiny to work with OnEntry.

Would you be open to me submitting a PR to add that functionality? Or do you have a plan to roll it into the wider changes you'd make to push, mentioned in #726?

For what it's worth, I agree with the proposed changes. It seems that people don't read the docs and fail to understand the differences between foreground & background notifications. I think the proposed change makes that much more explicit.

I completely understand & respect that Shiny isn't a product, it's code given away for free and I appreciate the effort that you put into it. That said, this is causing me a problem, and I'm keen to help resolve it if possible. Again, my apologies if this isn't the correct way to go about talking about this.

Thanks.

**Comment by aritchie** [comment]
@aspeckt112 This is the definitely the right place to do this.  

As far as a PR goes, the clickaction requirement is something done purely on the server.  It just has to point to the Shiny push notification receiver.

**Comment by aspeckt112** [comment]
Hey @aritchie 

Cheers for the reply above. I've had a pretty long read of the code, and I've decided to sponsor Shiny. I'm going to be using it for at least 3 apps in the next year. You've done really great work, the amount of hassle it saves is huge. In a couple of months, I should have some free time to contribute back. I'd be very happy to contribute to the docs / samples, whilst I continue learning the code.

Just to be 100% clear, and cement my understanding of this problem. When you've had the time to make the required changes, I'll simply need to specify "com.shiny.push.ShinyPushNotificationBroadcastReceiver" in my click_action? The issue right now is that the native adapter is only checking for the IntentKey "ShinyAndroidPush"?

**Comment by aritchie** [comment]
@aspeckt112 glad to hear Shiny will work for you!

If you specify the click action as "com.shiny.push.ShinyPushNotificationBroadcastReceiver.ENTRY_ACTION" now, it should cause the entry delegate function to fire within Shiny.  You can also set the click_action to any activity within android to run your own code.

**Comment by aspeckt112** [comment]
@aritchie 

Interesting, because I tried that, and all that happens is the app opens like normal, but hits the native adapter code and returns because there's no extra data.

I'll fully admit however, I'm a total Firebase beginner. The project I'm working on has been through multiple iterations and multiple developers, so I suspect there's some config issue going on, or the payload of the message from FCM is wrong.

I'll spin up a brand new sample project and try it tomorrow. I'm sure it'll work there, so it'll just be a case of going through the code I've inherited and trying to unpick it bit by bit. 

Thanks for your time, and I'm looking forward to diving into Shiny more.

**Comment by aritchie** [comment]
This is now fixed and will be available in preview builds now week.  Nothing special will be required from the server side.

The delegate and the info provided to the delegate may see the notification object removed.  This is still being debated.  You will still be given the payload data which is what most should be after anyhow.

**Comment by munkii** [comment]
I am still seeing an issue getting OnEntry to fire on the PushDelegate. I have read this issue including the later comments by ghost in spring 2022.

We have an ASP NET Core web API that calls the Azure Notification Hub but for my testing I've removed that piece and am just using the Test Send functionality in Azure Portal.

Based on the comments in Issue 656 my understanding is that I just need to set "click_action" in the notification node to "com.shiny.push.ShinyPushNotificationBroadcastReceiver.ENTRY_ACTION". Digging through the 2.7.0 branch of Shiny that class does not exist so I also tried "com.shiny.Notifications.ShinyNotificationBroadcastReceiver.ENTRY_ACTION". That did not work either.

The JSON message I was sending using Azure's Test Send was 

{
	"notification":{
		"title":"Notification Hub Test Notifications",
		"body":"This is a sample notification delivered by Azure Notification Hubs.",
                 "click_action": "com.shiny.push.ShinyPushNotificationBroadcastReceiver.ENTRY_ACTION"
	},
	"data":{
		"property1":"value1",
		"property2":42
	}
}

Whenever I set the click_action, the Notification arrives on the device but the app is not foregrounded when the notificarion is clicked and OnNewIntent override on MainActivity is not called. 

If I remove click_action from the payload OnNewIntent is called and the app is foregrounded but OnEntry is not called

I noticed in this [commit](https://github.com/shinyorg/shiny/commit/29fa28d9029aa0f82e2a37b810b5e2dd02734d05) (that is referenced in this issue) the NativeAdapter's TryProcessIntent mentions and intent extra of google.message_id. FWIW I am listing out intent.Extras.KeySet() and there is a google.message_id 0:1669730500406752%b5cef0c9b5cef0c9.

What am I missing here ?

I am using version 2.7.1 of 

- Shiny.Core
- Shiny.Logging.AppCenter
- Shiny.Notifications
- Shiny.Notifications.Common
- Shiny.PushDelegate

**Comment by aritchie** [comment]
@munkii 
The click_action has to be set to: SHINY_NOTIFICATION_CLICK
When you click on the notification, it will fire OnNewIntent in your activity.   The code you are referencing doesn't exist anymore.  

Please submit new issues with reproducible samples!

**Comment by MGohil** [comment]
@aritchie 

Unfortunately, I am seeing this issue in latest version 3.0.0-beta-0245 using .net maui in **Android** platform.

App is in **Foreground**:
OnReceived is fired when I send push from Azure push hub Test send.

App is in **Background**:
- No any event fires and I can't handle the push when I tap on push notification.
- I tried sending just "data" part without "notification" in payload, this fires OnReceived but no notification is displayed in tray.
- I tried passing action = "SHINY_PUSH_NOTIFICATION_CLICK" - this doesn't display notification in tray nor fires any event.

App is Not running (**Killed**):
- No any event fires and I can't handle the push when I tap on push notification.

I also tried pulling the Shiny repo and tested the same things in "Sample.Push.Maui" and same thing happens there.

Would you please guide me here what I am missing or doing wrong or if this is known issue in beta for Maui ?

Thanks
Milan G

**Comment by aritchie** [comment]
@MGohil Please open a new issue with reproducible sample and the push payload you are using.  If there is a bug, I will fix it.  If you've missed something in the repro, I will tell you.  I'm not here to guide, that is something I only do for sponsors at this point.  Code is open and free to view.
